ceph/ceph_admin/host.py

```diff
@@ -0,0 +1,386 @@
+"""Cephadm orchestration host operations."""
+import json
+import logging
+from copy import deepcopy
+
+from ceph.ceph import CephNode
+from ceph.utils import get_node_by_id
+
+from .common import config_dict_to_string
+from .orch import Orch, ResourceNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+class HostOpFailure(Exception):
+    pass
+
+
+class Host(Orch):
+    """Interface for executing ceph host <options> operations."""
+
+    SERVICE_NAME = "host"
+
+    def list(self):
+        """
+        List the cluster hosts
+        Returns:
+            json output of hosts list
+        """
+        return self.shell(
+            args=["ceph", "orch", "host", "ls", "--format=json"],
+        )
+
+    def add(self, config):
+        """
+        Add host to cluster
+        - add_label: host added with labels(roles assigned are considered)
+        - attach_address: host added with ip address(host ip address used)
+
+        Args:
+            config
+
+        config:
+            service: host
+            command: add
+            base_cmd_args:                          # arguments to ceph orch
+                concise: true
+                block: true
+            args:
+                node: "node2"                         # node-name or object
+                attach_ip_address: bool               # true or false
+                labels: [mon, osd] or apply-all-labels
+
+        labels are considered if list of strings are provided or all roles associated
+        with node will be considered if string "apply-all-labels"
+
+        """
+        cmd = ["ceph", "orch"]
+        if config.get("base_cmd_args"):
+            cmd.append(config_dict_to_string(config["base_cmd_args"]))
+
+        args = config["args"]
+        node = args.pop("node")
+        ceph_node = get_node_by_id(self.cluster, node_name=node)
+
+        if not ceph_node:
+            raise ResourceNotFoundError(f"No matching resource found: {node}")
+
+        attach_address = args.get("attach_ip_address")
+        _labels = args.get("labels")
+
+        if isinstance(_labels, str) and _labels == "apply-all-labels":
+            label_set = set(ceph_node.role.role_list)
+            _labels = list(label_set)
+
+        cmd.extend(["host", "add", ceph_node.shortname])
+
+        if attach_address:
+            cmd.append(ceph_node.ip_address)
+
+        if _labels:
+            # To fill mandate <address> argument, In case attach_address is False
```

Comment:
Shall we raise a usability issue in the upstream. Having positional arguments as empty does not make sense.